WIP mmu: enable virtual linking on x86

tyrantlizard77 opened the pull request:
This PR addresses remaining issues with a virtual kernel and enables it on a 32-bit x86 target to serve as a reference.

- Pull in changes in https://github.com/zephyrproject-rtos/zephyr/pull/29803
- Fixes mem_map test to not assume `test_page` is identity-mapped
- Restores virtual/physical conversion routines in mem_manage.h, but these still need work, they are perilous if not used properly and may need arch layer support if there are multiple physical RAM regions
- Update linker scripts such that LMA ranges represent physical address data will reside in at kernel start, and VMA ranges represent virtual address area the kernel will be linked at
- X86 linker script updates to support virtual linking
- Restore x86 code which supports dual linked page tables, instruction pointer transition at early boot, and proper handling of paging structure physical addresses. This still needs a little work, it makes the memory domain hacks for migrating threads even uglier so I will address this before this comes out of draft.

Commit:
qemu_x86: set machine type

Without this set, QEMU refuses to load the ELF image
if the entry address is a physical address outside
any linked sections. We need this as the kernel will
be linked at a virtual address but we still need to
start execution at the physical address where it is
loaded.

Signed-off-by: Andrew Boie <andrew.p.boie@intel.com>

Commit:
x86: 32-bit: set ENTRY properly

The entry point is at a physical memory location; use the
physical instead of virtual address of the start symbol.

Signed-off-by: Andrew Boie <andrew.p.boie@intel.com>

Commit:
linker-tool-gcc: revise for MMU support

We need to do a few things differently if we are to support
a virtual memory map, i.e. CONFIG_MMU where CONFIG_KERNEL_VM_BASE
is not the same as CONFIG_SRAM_BASE_ADDRESS.

 - All sections must be specified with a VMA and LMA, where
   VMA is the virtual address and LMA is the physical memory
   location.
 - All sections must be specified with ALIGN_WITH_INPUT to
   keep VMAs and LMAs synchronized

To do this, the existing linker macros need some adjustment:

 - GROUP_LINK_IN undefined on MMU platforms, and not used in
   any common linker scripts
 - New macro GROUP_ROM_LINK_IN for text/rodata sections
 - New macro GROUP_NOLOAD_LINK_IN for bss/noinit sections
 - Implicit ALIGN_WITH_INPUT for all sections

GROUP_FOLLOWS_AT is unused anywhere in the kernel for years
now and has been removed.

Signed-off-by: Andrew Boie <andrew.p.boie@intel.com>

Commit:
x86: 32-bit: restore virtual linking capability

This reverts commit 7d32e9f9a588fd3ed17277fb88210d453f2c9864.

We now allow the kernel to be linked virtually. This patch:

- Properly converts between virtual/physical addresses
- Handles early boot instruction pointer transition
- Double-maps SRAM to both virtual and physical locations
  in boot page tables to facilitate instruction pointer
  transition, with logic to clean this up after completed.

Signed-off-by: Andrew Boie <andrew.p.boie@intel.com>

Commit:
x86: set up linker ROM region for MMU

For virtual kernels this represents where the kernel
is loaded in physical memory.

Signed-off-by: Andrew Boie <andrew.p.boie@intel.com>

Commit:
tests: mem_map: use physical address of test page

It may not be identity-mapped.

Signed-off-by: Andrew Boie <andrew.p.boie@intel.com>

Commit:
WIP: kernel: mmu: restore addr conversion funcs

Currently needed by arch code and one test.

Signed-off-by: Andrew Boie <andrew.p.boie@intel.com>

Commit:
x86: add qemu_x86_virtual

Exercise in tests the kernel linked at a virtual base
address.

Signed-off-by: Andrew Boie <andrew.p.boie@intel.com>

andrewboie commented on include/linker/thread-local-storage.ld:
```diff
@@ -3,12 +3,12 @@
 	{
 		*(.tbss .tbss.* .gnu.linkonce.tb.* .tcommon);
-	} GROUP_LINK_IN(ROMABLE_REGION)
+	} GROUP_ROM_LINK_IN(RAMABLE_REGION, ROMABLE_REGION)
```
hmm this might need to be GROUP_NOLOAD_LINK_IN

Commit:
linker: tls: fix tdata/tbss locations

__tbss_start should be a virtual address.
__tdata_start should also be a virtual address, unless we're
using XIP in which case each thread should copy its thread-
local data out of flash.

Fixes issues with a kernel linked at a virtual address.

Signed-off-by: Andrew Boie <andrew.p.boie@intel.com>
